Add blog article: The case for Kubernetes limits (#43436)
kubernetes/website: content/en/blog/_posts/the-case-for-kubernetes-limits/index.md

Reviewer: Please change the `(1)` `(2)` and `(3)` to more obviously link - ideally, hyperlink - to the examples in the figure.

Author: I changed the text in the parentheses slightly to refer to the actual figure, but looks like I won't be able to do much more changes -- as far as I understand, `hugo` only generates `id` attributes for heading elements, and there's no way to add an `id` manually. The only chance would be to create a separate subsection for the figure, but that looks quite clumsy and would make things a bit less legible.

---
layout: blog
title: "The Case for Kubernetes Resource Limits: Predictability vs. Efficiency"

> date: TODO

Reviewer: Thanks. Please pick a November weekday later than the 10th of November, and use that date.

You should also include that date in the article filename.

Author: Done; the date chosen is 2023-11-16

> slug: the-case-for-kubernetes-resource-limits-predictability-vs-efficiency

Reviewer: I recommend:
```suggestion
slug: the-case-for-kubernetes-resource-limits
```

Most search engines etc can find the `<title>` and `<h1>` elements to learn the actual full title, and it's less to type.

---

> **Author:** Milan Plžík

Reviewer: ```suggestion
**Author:** Milan Plžík (Grafana Labs)
```

If you are not happy to blog as someone who works at Grafana, you could use “independent” instead.

Author: I'm definitely okay with that, I just didn't know the affiliation also goes to the Author field :)

_Milan Plzik is a Site Reliability Engineer on the [Grafana Labs](http://grafana.com) platform team, which maintains and improves internal infrastructure and tooling used by the product teams. As part of the infrastructure squad, he is primarily trying to make Kubernetes upgrades as smooth as possible. He likes going down the rabbit hole of various interesting Kubernetes issues._

> _Disclaimer: This article reflects the author's personal opinion; feel free to disagree. Also, there are a lot of subjective ideas that have not made it through any kind of rigorous testing. Use at your own risk._

Reviewer: We don't usually include either kind of a paragraph. If it's essential that we do, I'd prefer to boilerplate it so that all articles where its needed get the exact same text.

However, please do state or hint in the running text that these are your opinions and that different people in the end user community may have their won.

Author: Updated; PTAL.

> There’s been quite a lot of posts suggesting that not using Kubernetes resource limits might be a fairly useful thing (e.g., [here](https://home.robusta.dev/blog/stop-using-cpu-limits/) or [here](https://erickhun.com/posts/kubernetes-faster-services-no-cpu-limits/) ). The points made there are totally valid – it doesn’t make much sense to pay for compute power that will not be used due to limits, nor to artificially increase latency. This post strives to argue that limits have their legitimate use as well.

Reviewer: Please avoid “here” in hyperlinks, even for blog articles. See https://kubernetes.io/docs/contribute/style/style-guide/ - much of which actually doesn't apply to blog writing.

Author: Fixed; replaced with actual article names.

Now, with Black Friday and Cyber Monday approaching, businesses expect a surge in traffic. Good performance data/benchmarks of the past performance allow businesses to plan for some extra capacity. But is data about pods without limits reliable? With memory or CPU instant spikes handled by the extra resources, everything might look good according to past data. But once the pod bin-packing changes and the extra resources get more scarce, everything might start looking different – ranging from request latencies rising negligibly to requests slowly snowballing and causing pod OOM kills. While almost no one actually cares about the former, the latter is a serious issue that requires instant capacity increase.

## Configuring the limits
Not using limits takes a tradeoff – it opportunistically improves the performance if there are extra resources available, but lowers predictability of the performance, which might strike back in the future. There are a few approaches that can be used to increase the predictability again. Let’s pick two of them to analyze:

> - *Configure workload limits to be a fixed (and small) percentage more than the requests – let’s call it “fixed-fraction headroom".* This allows the use of some extra shared resources, but keeps the per-node overcommit bound and can be taken to guide worst-case estimates for the workload. Note that the bigger the limits percentage is, the bigger the variance in the performance that might happen across the workloads.
> - *Configure workloads with requests = limits.* From some point of view, this is equivalent to giving each pod its own tiny machine with constrained resources; the performance is fairly predictable. This also puts the pod into the Guaranteed QoS class, which makes it get evicted only after Best Effort and Burstable pods have been evicted by a node under resource pressure (see [here](https://kubernetes.io/docs/concepts/workloads/pods/pod-qos/)).

Reviewer: ```suggestion
- **Configure workload limits to be a fixed (and small) percentage more than the requests** – I'll call it _fixed-fraction headroom). This allows the use of some extra shared resources, but keeps the per-node overcommit bound and can be taken to guide worst-case estimates for the workload. Note that the bigger the limits percentage is, the bigger the variance in the performance that might happen across the workloads.
- **Configure workloads with `requests` = `limits`**. From some point of view, this is equivalent to giving each pod its own tiny machine with constrained resources; the performance is fairly predictable. This also puts the pod into the _Guaranteed_ QoS class, which makes it get evicted only after _BestEffort_ and _Burstable) pods have been evicted by a node under resource pressure (see [Quality of Service for Pods](/docs/concepts/workloads/pods/pod-qos/)).
```
If you want to, the first use of each QoS class name could be a link to the explanation of that class.

Some other cases might also be considered, but these are probably the two simplest ones to discuss.

> 

Reviewer: Which resources should people set requests and limits for? Just CPU and memory? Anything else that is sometimes worth defining?

Author: It's debatable, but I think CPU and memory is okay-ish here.

IIUC, the only resources that can be requested/limited by Kubernetes API are CPU and memory (and possibly hugepages, ephemeral-storage): https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-requests-and-limits-of-pod-and-container . I believe the former two cover most of the use-cases and is much easier to imagine than the latter two, so I'd rather focus on those.

As for defining more things -- I'd like things lean and not introduce additional strategies that might be hard to evaluate/do estimates on.

Note that in both cases discussed above, we’re effectively preventing the workloads from using some cluster resources it has at the cost of getting more predictability – which might sound like a steep price to pay for a bit more stable performance. Let’s try to quantify the impact there.

### Bin-packing and cluster resource allocation
Firstly, let’s discuss bin-packing and cluster resource allocation. There’s some inherent cluster inefficiency that comes to play – it’s hard to achieve 100% resource allocation in a Kubernetes cluster. Thus, some percentage will be left unallocated.

When configuring fixed-fraction headroom limits, a proportional amount of this will be available to the pods. If the percentage of unallocated resources in the cluster is lower than the constant we use for setting fixed-fraction headroom limits (2), all the pods together are able to theoretically use up all the node’s resources; otherwise there are some resources that will inevitably be wasted (1). In order to eliminate the inevitable resource waste, the percentage for fixed-fraction headroom limits should be configured so that it’s at least equal to the expected percentage of unallocated resources.

> <img alt="Chart displaying various requests/limits configurations" width="40%" src="requests-limits-configurations.svg">

Reviewer: Please use a _figure_ - see https://kubernetes.io/docs/contribute/style/diagram-guide/ for hints.

Author: Replaced with the Hugo `figure` shortcode.
